### PR TITLE
Ep/master build

### PR DESCRIPTION
**Pull request by eliasjpr**
More details to come...

> Tests are running but currently failing.

Depends on 
https://github.com/amberframework/amber-spec/pull/12
https://github.com/amberframework/granite-orm/pull/89

### CODE REVIEW

File: spec/amber/validations/params_spec.cr [hunk 1]
```diff
@@ -5,7 +5,7 @@ module Amber::Validators
     describe "#apply" do
       it "raises error when missing field" do
         params = params_builder("")
-        rule = BaseRule.new("field", "") do
+        rule = BaseRule.new("field", :string, "") do
```
Comment (marksiemers) [inline spec/amber/validations/params_spec.cr]: I got this error while trying to run the spec:
```
Error in line 34: while requiring "./spec/amber/validations/params_spec.cr"

in spec/amber/validations/params_spec.cr:8: wrong number of arguments for 'Amber::Validators::BaseRule.new' (given 3, expected 2)
Overloads are:
 - Amber::Validators::BaseRule.new(field : String | Symbol, msg : String | ::Nil, &block)

        rule = BaseRule.new("field", :string, "") do
```

File: src/amber/cli/templates/scaffold/controller/crecto/spec/controllers/{{name}}_controller_spec.cr.ecr [hunk 2]
```diff
@@ -1,71 +1,86 @@
...
 
-describe <%= class_name %>Controller do
+class <%= class_name %>ControllerTest < Spec::Controller::Test
+  getter handler : Amber::Pipe::Pipeline
+
+  def initialize
+    @handler = Amber::Server.settings.handler
+    @handler.prepare_pipelines
+  end
+end
```
Comment (marksiemers) [inline src/amber/cli/templates/scaffold/controller/crecto/spec/controllers/{{name}}_controller_spec.cr.ecr]: Is this working? That's great if it does. What about CSRF? Is that defaulted to be part of `Amber::Server.settings.handler`?

File: src/amber/cli/templates/scaffold/controller/crecto/spec/controllers/{{name}}_controller_spec.cr.ecr [hunk 3]
```diff
@@ -1,71 +1,86 @@
...
     <%= class_name %>.clear
   end
 
   describe "index" do
     it "renders all the <%= @name %>s" do
-      subject = create_subject
-      get "/<%= @name %>s"
+      model = create_model
+      response = subject.get "/<%= @name %>s"
       response.body.should contain "test"
```
Comment (marksiemers) [inline src/amber/cli/templates/scaffold/controller/crecto/spec/controllers/{{name}}_controller_spec.cr.ecr]: If the first field is a reference, then this will fail. Is there a different expectation that we can create? What the response body contains could be considered a view or behavior test.

The controller should return a 200 status code, a non-empty body, and I'm not sure what else.

File: src/amber/validations/params.cr [hunk 4]
```diff
@@ -122,4 +122,4 @@ module Amber::Validators
       @rules << rule
     end
   end
-end
+end
```
Comment (marksiemers) [inline src/amber/validations/params.cr]: Do we have a guideline on this? I like having a newline at the end of every file.

File: src/amber/cli/templates/scaffold/controller/granite/spec/controllers/{{name}}_controller_spec.cr.ecr [hunk 5]
```diff
@@ -1,71 +1,86 @@
 <% field_name = @fields.first.name -%>
```
Comment (marksiemers) [inline src/amber/cli/templates/scaffold/controller/granite/spec/controllers/{{name}}_controller_spec.cr.ecr]: Are the controllers and the tests any different between granite and crecto? Is there a way to structure these files so we don't have the same tests all written twice?

File: spec/amber/cli/build_spec.cr [hunk 6]
```diff
@@ -0,0 +1,62 @@
+require "../../spec_helper"
```
Comment (marksiemers) [inline spec/amber/cli/build_spec.cr]: The check for the `run_build_tests` flag is gone - so this is running by default. Is that intended?

File: spec/amber/cli/build_spec.cr [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+      MainCommand.run ["generate", "socket", "Eagle"] | ["soar", "nest"]
+      MainCommand.run ["generate", "channel", "Falcon"]
+
+      prepare_yaml(Dir.current)
+
+      puts "RUNNING: shard build #{TESTING_APP} - started..."
+      build_result = `shards build #{TEST_APP_NAME}`
+      puts "#{TESTING_APP} build completed..."
+
+      sleep 3000
```
Comment (marksiemers) [inline spec/amber/cli/build_spec.cr]: I think in crystal the sleep statements are in seconds, not milliseconds.

File: src/amber/cli/templates/scaffold/controller/crecto/spec/controllers/{{name}}_controller_spec.cr.ecr [hunk 8]
```diff
@@ -1,71 +1,86 @@
...
+  model.<%= field_name %> = <%= field_name.camelcase %>.new 
+  <% else -%>
+  model.<%= field_name %> = "test"
+  <% end -%>
+  model.save
+  model
 end
 
-describe <%= class_name %>Controller do
+class <%= class_name %>ControllerTest < Spec::Controller::Test
```
Comment (marksiemers) [inline src/amber/cli/templates/scaffold/controller/crecto/spec/controllers/{{name}}_controller_spec.cr.ecr]: I'm getting an error, do we need to release a new version of `amber-spec` and have the generated app pick up that version? Or for testing tell it to pull from master, or a certain commit hash?
`undefined constant Spec::Controller::Test`
```
.====== START Creating database for test_app ======
====== DONE Database created test_app ======
FF.cleaning up...


Failures:

  1) building a generated app crystal spec can be executed
     Failure/Error: spec_result.should contain "Finished in"

       Expected:   "Error in line 1: while requiring \"./spec/controllers/animal_controller_spec.cr\"\n\nin spec/controllers/animal_controller_spec.cr:10: \e[1mundefined constant Spec::Controller::Test\e[0m\n\nclass AnimalControllerTest < Spec::Controller::Test\n                             \e[32;1m^~~~~~~~~~~~~~~~~~~~~~\e[0m\n\n"
       to include: "Finished in"

     # spec/amber/cli/build_spec.cr:47

  2) building a generated app crystal spec has no errors
     Failure/Error: spec_result.should_not contain "Error in line"

       Expected: value "Error in line 1: while requiring \"./spec/controllers/animal_controller_spec.cr\"\n\nin spec/controllers/animal_controller_spec.cr:10: \e[1mundefined constant Spec::Controller::Test\e[0m\n\nclass AnimalControllerTest < Spec::Controller::Test\n                             \e[32;1m^~~~~~~~~~~~~~~~~~~~~~\e[0m\n\n"
       to not include: "Error in line"

     # spec/amber/cli/build_spec.cr:51

Finished in 1:39 minutes
30 examples, 2 failures, 0 errors, 0 pending

Failed examples:

crystal spec spec/amber/cli/build_spec.cr:46 # building a generated app crystal spec can be executed
crystal spec spec/amber/cli/build_spec.cr:50 # building a generated app crystal spec has no errors
```

File: spec/amber/cli/build_spec.cr (hunk 6 above)
Comment (eliasjpr) [reply spec/amber/cli/build_spec.cr]: I am planning to add the flag when I am done I had to keep reminding myself to run the spec with the flag :)

File: spec/amber/cli/build_spec.cr (hunk 7 above)
Comment (eliasjpr) [reply spec/amber/cli/build_spec.cr]: Yes this is intentional while I work on testing the generated app it gives me time too manually run the specs and checks things out

Comment (marksiemers) [reply spec/amber/cli/build_spec.cr]: I see. You could probably remove the call to `cleanup` and the app will stick around forever. (until you rm -rf it)

File: src/amber/cli/templates/scaffold/controller/crecto/spec/controllers/{{name}}_controller_spec.cr.ecr [hunk 9]
```diff
@@ -1,5 +1,6 @@
 <% field_name = @fields.first.name -%>
 <% human_name = @name.split("_").map(&.capitalize).join(" ") %>
+<% create_method_name = "create_#{@name.downcase}" %>
```
Comment (marksiemers) [inline src/amber/cli/templates/scaffold/controller/crecto/spec/controllers/{{name}}_controller_spec.cr.ecr]: This likely does no harm, but the name should already be `underscore`d

In general for the templates - for the names, we need to define all of them as instance methods in the class itself so it is consistent.

The full convention for naming will probably continue to evolve - subing out all characters that aren't alphanumeric or underscore, etc.

Then there is the whole pluralization, but I digress - for a different day.

File: spec/amber/cli/build_spec.cr [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+      MainCommand.run ["generate", "scaffold", "PostComment"] | (options + ["post:reference"])
+      MainCommand.run ["generate", "model", "Bat"] | options
+      MainCommand.run ["generate", "migration", "Crocodile"] | options
+      MainCommand.run ["generate", "mailer", "Dinosaur"] | options
+      MainCommand.run ["generate", "socket", "Eagle"] | ["soar", "nest"]
+      MainCommand.run ["generate", "channel", "Falcon"]
+
+      prepare_yaml(Dir.current)
+
+      prepare_db_yml if ENV["CI"]? == "true"
```
Comment (eliasjpr) [inline spec/amber/cli/build_spec.cr]: This is so we can run the test in CI and Locally with no manual changes

File: spec/support/helpers/cli_helper.cr [hunk 11]
```diff
@@ -57,6 +57,13 @@ module CLIHelper
     File.write("#{path}/shard.yml", shard)
   end
 
+  def prepare_db_yml
```
Comment (eliasjpr) [inline spec/support/helpers/cli_helper.cr]: This is so we can run the test in CI and Locally with no manual changes.

File: src/amber/cli/templates/scaffold/controller.cr [hunk 12]
```diff
@@ -49,5 +60,22 @@ module Amber::CLI::Scaffold
         f.reference? ? "#{f.name}_id" : f.name
       end
     end
+
+    private def default_value(field_type)
```
Comment (eliasjpr) [inline src/amber/cli/templates/scaffold/controller.cr]: This is to assign default values for the generated specs

File: src/amber/server/configuration.cr [hunk 13]
```diff
@@ -44,6 +44,10 @@ module Amber
         settings.router
       end
 
+      def self.handler
```
Comment (eliasjpr) [inline src/amber/server/configuration.cr]: Convenience method to write `Amber::Server.handler` instead `Amber::Server.settings.handler`

File: src/amber/cli/templates/migration.cr [hunk 14]
```diff
@@ -14,12 +14,12 @@ module Amber::CLI
     def initialize(@name, fields)
       @fields = fields.map { |field| Field.new(field) }
       @database = database
-      @timestamp = Time.now.to_s("%Y%m%d%H%M%S")
+      @timestamp = Time.now.to_s("%Y%m%d%H%M%S%L")
```
Comment (eliasjpr) [inline src/amber/cli/templates/migration.cr]: Ran into an issue where the migrations had the same timestamps because crystal was so fast to generate them. All test files generated within the same second. `%L` adds milliseconds, cause thats how we roll!

File: src/amber/cli/templates/app/shard.yml.ecr [hunk 15]
```diff
@@ -51,8 +51,8 @@ dependencies:
     github: crystal-lang/crystal-sqlite3
     version: ~> 0.8.3
 <% end -%>
- 
+
 development_dependencies:
-  amber_spec:
-    github: amberframework/amber-spec
-    version: ~> 0.1.2
+  garnet_spec:
```
Comment (eliasjpr) [inline src/amber/cli/templates/app/shard.yml.ecr]: Garnet!!!

File: src/amber/cli/templates/app/config/environments/development.yml.ecr [hunk 16]
```diff
@@ -20,7 +20,7 @@ secrets:
   when "mysql" -%>
   database: mysql://root@localhost:3306/<%= database_name_base %>_development
 <%when "pg" -%>
-  database: postgres://postgres:@pg:5432/<%= database_name_base %>_development
+  database: postgres://postgres:@localhost:5432/<%= database_name_base %>_development
```
Comment (eliasjpr) [inline src/amber/cli/templates/app/config/environments/development.yml.ecr]: These were not consistent with the current `config/database.yml`. Two things to note.

- We should get rid of `config/database.yml`
- Be able to load environment dynamically `Amber.env.load(:test)`

File: spec/amber/cli/build_spec.cr (hunk 6 above)
Comment (eliasjpr) [reply spec/amber/cli/build_spec.cr]: 🎏  Are up!

File: src/amber/server/configuration.cr (hunk 13 above)
Comment (elorest) [reply src/amber/server/configuration.cr]: I don't think the base handler should be included in settings to begin with. It's not a setting as the entire routing system depends on it.

Comment (elorest) [reply src/amber/server/configuration.cr]: Same with pubsub_adapter and router. WDYT? It's not what come to mind for me when I think of settings.

Comment (eliasjpr) [reply src/amber/server/configuration.cr]: Agreed. See https://github.com/amberframework/amber/pull/386

File: spec/amber/cli/build_spec.cr [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+          spec_result.should contain "Finished in"
+        end
+
+        it "has no errors" do
+          spec_result.should_not contain "Error in line"
+        end
+
+        it "has no failures" do
+          spec_result.should_not contain "Failures"
+        end
```
Comment (marksiemers) [inline spec/amber/cli/build_spec.cr]: For the 3 preceding tests, should we keep lines like below to help us debug (even on travis):
```crystal
it "has no failures" do
  puts spec_result if spec_result.includes? "Failures"
  spec_result.should_not contain "Failures"
end

File: src/amber/cli/templates/scaffold/controller/crecto/spec/controllers/{{name}}_controller_spec.cr.ecr [hunk 18]
```diff
@@ -1,72 +1,110 @@
-<% field_name = @fields.first.name -%>
+<% params_name =  "#{@name.downcase}_params" -%>
+<% hash_name   =  "#{@name.downcase}_hash" -%>
+<% model_name  =  "create_#{@name.downcase}" -%>
```
Comment (marksiemers) [inline src/amber/cli/templates/scaffold/controller/crecto/spec/controllers/{{name}}_controller_spec.cr.ecr]: It shouldn't be necessary to modify the name, but if we do, use `@name.underscore` to be safe.

File: src/amber/cli/templates/scaffold/controller/crecto/spec/controllers/{{name}}_controller_spec.cr.ecr [hunk 19]
```diff
@@ -1,72 +1,110 @@
...
-  end
+def <%= params_name %>
+  params = [] of String
+  <%- @fields_hash.keys.map do |k| -%>
+  params << "<%= k %>=#{<%= hash_name %>["<%= k %>"]}"
+  <%- end -%>
+  params.join("&")
+end
+
+def <%= model_name %>
```
Comment (marksiemers) [inline src/amber/cli/templates/scaffold/controller/crecto/spec/controllers/{{name}}_controller_spec.cr.ecr]: `create_model_method` is more descriptive for what this is.

File: src/amber/cli/templates/scaffold/controller/crecto/spec/controllers/{{name}}_controller_spec.cr.ecr [hunk 20]
```diff
@@ -1,72 +1,110 @@
...
+    @handler = Amber::Pipe::Pipeline.new
+    @handler.build :web do
+      plug Amber::Pipe::Error.new
+      plug Amber::Pipe::Logger.new
+      plug Amber::Pipe::Session.new
+      plug Amber::Pipe::Flash.new
     end
+    @handler.prepare_pipelines
   end
+end
```
Comment (marksiemers) [inline src/amber/cli/templates/scaffold/controller/crecto/spec/controllers/{{name}}_controller_spec.cr.ecr]: This mock looks good for this release, but I want to revisit in the future. The controller tests in an app should test the actual controller (and actual pipeline) that the app is using. Even if that means figuring out CSRF.

The mock used in the garnet spec tests is just because there is no actual web app to test (in that repo).

File: src/amber/cli/templates/scaffold/controller/granite/spec/controllers/{{name}}_controller_spec.cr.ecr [hunk 21]
```diff
@@ -1,72 +1,110 @@
-<% field_name = @fields.first.name -%>
+<% params_name =  "#{@name.downcase}_params" -%>
+<% hash_name   =  "#{@name.downcase}_hash" -%>
+<% model_name  =  "create_#{@name.downcase}" -%>
```
Comment (marksiemers) [inline src/amber/cli/templates/scaffold/controller/granite/spec/controllers/{{name}}_controller_spec.cr.ecr]: Same thing here, shouldn't require anything, but use `underscore` if anything.

File: src/amber/cli/templates/scaffold/controller/granite/spec/controllers/{{name}}_controller_spec.cr.ecr [hunk 22]
```diff
@@ -1,72 +1,110 @@
...
-  end
+def <%= params_name %>
+  params = [] of String
+  <%- @fields_hash.keys.map do |k| -%>
+  params << "<%= k %>=#{<%= hash_name %>["<%= k %>"]}"
+  <%- end -%>
+  params.join("&")
+end
+
+def <%= model_name %>
```
Comment (marksiemers) [inline src/amber/cli/templates/scaffold/controller/granite/spec/controllers/{{name}}_controller_spec.cr.ecr]: Same comment `create_model_method` is more descriptive.

File: src/amber/cli/templates/scaffold/controller/granite/spec/controllers/{{name}}_controller_spec.cr.ecr [hunk 23]
```diff
@@ -1,72 +1,110 @@
...
+    @handler = Amber::Pipe::Pipeline.new
+    @handler.build :web do
+      plug Amber::Pipe::Error.new
+      plug Amber::Pipe::Logger.new
+      plug Amber::Pipe::Session.new
+      plug Amber::Pipe::Flash.new
     end
+    @handler.prepare_pipelines
   end
+end
```
Comment (marksiemers) [inline src/amber/cli/templates/scaffold/controller/granite/spec/controllers/{{name}}_controller_spec.cr.ecr]: Same comment, we should revisit this for 0.4 to test the app's controller.

File: src/amber/cli/templates/scaffold/view/src/views/{{name}}/show.slang.ecr [hunk 24]
```diff
@@ -1,3 +1,4 @@
+<h1>Show <%= display_name %></h1>
```
Comment (marksiemers) [inline src/amber/cli/templates/scaffold/view/src/views/{{name}}/show.slang.ecr]: For slang, this should be `h1 Show <%= display_name %>`

**Comment by eliasjpr** [comment]
Definitely this mock should be abstracted out of the controller test, and we should provide a better way to redefining the pipeline if this is what we want to allow.

Given that controller tests are unit tests we need a way to use render so we can assert against. I am not concern of testing the pipeline here since that should be tested on its own, pipelines are at a higher level and should and testing those here will mean that controller tests will be mixed with integration tests. The controller is only affecting the response, the important bits that a controller should be testing are (I could be missing some other):

- Rendered templates
- Redirects
- Instance variables assigned in the controller to be shared with the view
- Cookies sent back with the response

Basically anything that the controllers are changing that will affect the response

**Comment by marksiemers** [comment]
@eliasjpr - You make good points RE: controller tests. I think we should further discuss the design of the generated test suite for a later release.

For this one, it is fine as-is (using the mock), if the other small changes can be made (or discussed), we can get this merged in.

**Comment by marksiemers** [comment]
@elorest @drujensen - Review and approve?

**Comment by marksiemers** [comment]
@elorest - A number of these commits are mine, so I'd feel more comfortable with a 3rd approval.

Then, unless @eliasjpr has any objection, we can squash and merge.
